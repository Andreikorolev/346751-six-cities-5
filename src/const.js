export const Cities = [
  `Paris`,
  `Cologne`,
  `Brussels`,
  `Amsterdam`,
  `Hamburg`,
  `Dusseldorf`,
];

export const SortOptions = {
  popular: `Popular`,
  lowToHigh: `low to high`,
  highToLow: `high to low`,
  rating: `Top rated first`,
};
