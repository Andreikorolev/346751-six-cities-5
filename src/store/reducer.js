import {extend} from "../utils.js";
import {ActionType} from "./action";
import offers from "../mocs/offers";
import reviews from "../mocs/reviews";
import {Cities} from "../const";

const initialState = {
  city: Cities[0],
  offers: offers[Cities[0]],
  reviews,
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.CITY_CHANGE:
      return extend(state, {
        city: action.payload,
      });

    case ActionType.GET_OFFERS_LIST:
      return extend(state, {
        offers: offers[action.payload],
      });
  }

  return state;
};

export {reducer};
