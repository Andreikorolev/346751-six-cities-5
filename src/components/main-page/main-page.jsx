import React from "react";
import PropTypes from "prop-types";
import {connect} from "react-redux";
import {ActionCreator} from "../../store/action";

import Header from "../header/header";
import Main from "../main/main";
import CitiesList from "../cities-list/cities-list";
import OffersList from "../offers-list/offers-list";
import OffersMap from "../offers-map/offers-map";
import SortingOptions from "../sorting-options/sorting-options";


const MainPage = ({offers, city, onCityChange, sortOption, onSortingOptionChange}) => {

  return (
    <div className="page page--gray page--main">
      <Header/>
      <Main renderClassName={() => (`page__main--index`)}>
        <h1 className="visually-hidden">Cities</h1>
        <CitiesList
          city={city}
          onCityChange={onCityChange}
        />
        <div className="cities">
          <div className="cities__places-container container">
            <section className="cities__places places">
              <h2 className="visually-hidden">Places</h2>
              <b className="places__found">{(offers.length)} places to stay in {city}</b>
              <SortingOptions
                sortOption={sortOption}
                offers={offers}
                onSortingOptionChange={onSortingOptionChange}
              />
              <OffersList
                offers={offers}
                renderClassName={() => (`cities__places-list tabs__content`)}
                renderOfferMark={() => (true)}
              />
            </section>
            <div className="cities__right-section">
              <OffersMap
                offers={offers}
                activeZoomControl={false}
                activeScrollWheelZoom={true}
              />
            </div>
          </div>
        </div>
      </Main>
    </div>
  );
};


MainPage.propTypes = {
  city: PropTypes.string.isRequired,
  onCityChange: PropTypes.func.isRequired,
  offers: PropTypes.array.isRequired,
  sortOption: PropTypes.string.isRequired,
  onSortingOptionChange: PropTypes.func.isRequired,
};

const mapStateToProps = (state) => ({
  city: state.city,
  offers: state.offers,
  sortOption: state.sortOption,
});

const mapDispatchToProps = (dispatch) => ({
  onCityChange(item) {
    dispatch(ActionCreator.cityChange(item));
    dispatch(ActionCreator.getOffersList(item));
  },
  onSortingOptionChange(item, offers = offers) {
    dispatch(ActionCreator.sortingOptionChange(item, offers = offers));
  }
});

export {MainPage};
export default connect(mapStateToProps, mapDispatchToProps)(MainPage);
