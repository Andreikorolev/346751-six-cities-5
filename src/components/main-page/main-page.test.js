import React from "react";
import renderer from "react-test-renderer";
import {MainPage} from "./main-page";
import {BrowserRouter} from "react-router-dom";
import {Provider} from "react-redux";
import {createStore} from "redux";
import reducer from "../../store/reducers/root-reducer";
import {offers} from "../../../mocks/mocks";
import {SortOptions} from "../../const";

const noop = () => {};
const store = createStore(reducer);
jest.mock(`../offers-map/offers-map`, () => `Map`);

describe(`MainPage render`, () => {
  it(`should render without offers`, () => {
    const tree = renderer
      .create(
          <Provider store={store}>
            <BrowserRouter>
              <MainPage
                city={`Paris`}
                onCityChange={noop}
                offers={[]}
                sortOption={SortOptions.POPULAR}
              />
            </BrowserRouter>
          </Provider>
      ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`should render with offers`, () => {
    const tree = renderer
      .create(
          <Provider store={store}>
            <BrowserRouter>
              <MainPage
                city={`Paris`}
                onCityChange={noop}
                offers={offers}
                sortOption={SortOptions.LOW_TO_HIGH}
              />
            </BrowserRouter>
          </Provider>
      ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
