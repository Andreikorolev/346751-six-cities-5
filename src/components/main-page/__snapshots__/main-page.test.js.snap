// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainPage render should render with offers 1`] = `
<div
  className="page page--gray page--main"
>
  <header
    className="header"
  >
    <div
      className="container"
    >
      <div
        className="header__wrapper"
      >
        <div
          className="header__left"
        >
          <a
            className="header__logo-link"
            href="/"
            onClick={[Function]}
          >
            <img
              alt="6 cities logo"
              className="header__logo"
              height="41"
              src="img/logo.svg"
              width="81"
            />
          </a>
        </div>
        <nav
          className="header__nav"
        >
          <ul
            className="header__nav-list"
          >
            <li
              className="header__nav-item user"
            >
              <a
                className="header__nav-link header__nav-link--profile"
                href="/login"
                onClick={[Function]}
              >
                <div
                  className="header__avatar-wrapper user__avatar-wrapper"
                />
                <span
                  className="header__user-name user__name"
                >
                  Sing in
                </span>
              </a>
            </li>
          </ul>
        </nav>
      </div>
    </div>
  </header>
  <main
    className="page__main undefined"
  >
    <h1
      className="visually-hidden"
    >
      Cities
    </h1>
    <div
      className="tabs"
    >
      <section
        className="locations container"
      >
        <ul
          className="locations__list tabs__list"
        >
          <li
            className="locations__item"
            onClick={[Function]}
          >
            <a
              className="locations__item-link tabs__item tabs__item--active"
            >
              <span>
                Paris
              </span>
            </a>
          </li>
          <li
            className="locations__item"
            onClick={[Function]}
          >
            <a
              className="locations__item-link tabs__item"
            >
              <span>
                Cologne
              </span>
            </a>
          </li>
          <li
            className="locations__item"
            onClick={[Function]}
          >
            <a
              className="locations__item-link tabs__item"
            >
              <span>
                Brussels
              </span>
            </a>
          </li>
          <li
            className="locations__item"
            onClick={[Function]}
          >
            <a
              className="locations__item-link tabs__item"
            >
              <span>
                Amsterdam
              </span>
            </a>
          </li>
          <li
            className="locations__item"
            onClick={[Function]}
          >
            <a
              className="locations__item-link tabs__item"
            >
              <span>
                Hamburg
              </span>
            </a>
          </li>
          <li
            className="locations__item"
            onClick={[Function]}
          >
            <a
              className="locations__item-link tabs__item"
            >
              <span>
                Dusseldorf
              </span>
            </a>
          </li>
        </ul>
      </section>
    </div>
    <div
      className="cities"
    >
      <div
        className="cities__places-container container"
      >
        <section
          className="cities__places places"
        >
          <h2
            className="visually-hidden"
          >
            Places
          </h2>
          <b
            className="places__found"
          >
            4
             places to stay in 
            Paris
          </b>
          <form
            action="#"
            className="places__sorting"
            method="get"
          >
            <span
              className="places__sorting-caption"
            >
              Sort by 
            </span>
            <span
              className="places__sorting-type"
              onClick={[Function]}
              tabIndex="0"
            >
              Popular
              <svg
                className="places__sorting-arrow"
                height="4"
                width="7"
              >
                <use
                  xlinkHref="#icon-arrow-select"
                />
              </svg>
            </span>
            <ul
              className="places__options places__options--custom  "
            >
              <li
                className="places__option places__option--active"
                onClick={[Function]}
                tabIndex="0"
              >
                Popular
              </li>
              <li
                className="places__option "
                onClick={[Function]}
                tabIndex="0"
              >
                low to high
              </li>
              <li
                className="places__option "
                onClick={[Function]}
                tabIndex="0"
              >
                high to low
              </li>
              <li
                className="places__option "
                onClick={[Function]}
                tabIndex="0"
              >
                Top rated first
              </li>
            </ul>
          </form>
          <div
            className="cities__places-list tabs__content places__list"
          >
            <article
              className=" cities__place-card place-card"
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
            >
              <div
                className="place-card__mark"
              >
                <span>
                  Premium
                </span>
              </div>
              <div
                className="cities__image-wrapper place-card__image-wrapper"
              >
                <a
                  href="/offer/4"
                  onClick={[Function]}
                >
                  <img
                    alt="Place image"
                    className="place-card__image"
                    height="200"
                    src="img/1.png"
                    width="260"
                  />
                </a>
              </div>
              <div
                className="cities__card-info place-card__info"
              >
                <div
                  className="place-card__price-wrapper"
                >
                  <div
                    className="place-card__price"
                  >
                    <b
                      className="place-card__price-value"
                    >
                      €
                      120
                    </b>
                    <span
                      className="place-card__price-text"
                    >
                      / night
                    </span>
                  </div>
                  <button
                    className="place-card__bookmark-button button place-card__bookmark-button--active"
                    onClick={[Function]}
                    type="button"
                  >
                    <svg
                      className="place-card__bookmark-icon"
                      height="19"
                      width="18"
                    >
                      <use
                        xlinkHref="#icon-bookmark"
                      />
                    </svg>
                    <span
                      className="visually-hidden"
                    >
                      To bookmarks
                    </span>
                  </button>
                </div>
                <div
                  className="place-card__rating rating"
                >
                  <div
                    className="place-card__stars rating__stars"
                  >
                    <span
                      style={
                        Object {
                          "width": "60%",
                        }
                      }
                    />
                    <span
                      className="visually-hidden"
                    >
                      Rating
                    </span>
                  </div>
                </div>
                <h2
                  className="place-card__name"
                >
                  <a
                    href="/offer/4"
                    onClick={[Function]}
                  >
                    Beautiful & luxurious studio at great location
                  </a>
                </h2>
                <p
                  className="place-card__type"
                >
                  apartment
                </p>
              </div>
            </article>
            <article
              className=" cities__place-card place-card"
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
            >
              <div
                className="cities__image-wrapper place-card__image-wrapper"
              >
                <a
                  href="/offer/10"
                  onClick={[Function]}
                >
                  <img
                    alt="Place image"
                    className="place-card__image"
                    height="200"
                    src="img/1.png"
                    width="260"
                  />
                </a>
              </div>
              <div
                className="cities__card-info place-card__info"
              >
                <div
                  className="place-card__price-wrapper"
                >
                  <div
                    className="place-card__price"
                  >
                    <b
                      className="place-card__price-value"
                    >
                      €
                      120
                    </b>
                    <span
                      className="place-card__price-text"
                    >
                      / night
                    </span>
                  </div>
                  <button
                    className="place-card__bookmark-button button"
                    onClick={[Function]}
                    type="button"
                  >
                    <svg
                      className="place-card__bookmark-icon"
                      height="19"
                      width="18"
                    >
                      <use
                        xlinkHref="#icon-bookmark"
                      />
                    </svg>
                    <span
                      className="visually-hidden"
                    >
                      To bookmarks
                    </span>
                  </button>
                </div>
                <div
                  className="place-card__rating rating"
                >
                  <div
                    className="place-card__stars rating__stars"
                  >
                    <span
                      style={
                        Object {
                          "width": "40%",
                        }
                      }
                    />
                    <span
                      className="visually-hidden"
                    >
                      Rating
                    </span>
                  </div>
                </div>
                <h2
                  className="place-card__name"
                >
                  <a
                    href="/offer/10"
                    onClick={[Function]}
                  >
                    Beautiful & luxurious studio at great location
                  </a>
                </h2>
                <p
                  className="place-card__type"
                >
                  apartment
                </p>
              </div>
            </article>
            <article
              className=" cities__place-card place-card"
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
            >
              <div
                className="place-card__mark"
              >
                <span>
                  Premium
                </span>
              </div>
              <div
                className="cities__image-wrapper place-card__image-wrapper"
              >
                <a
                  href="/offer/2"
                  onClick={[Function]}
                >
                  <img
                    alt="Place image"
                    className="place-card__image"
                    height="200"
                    src="img/1.png"
                    width="260"
                  />
                </a>
              </div>
              <div
                className="cities__card-info place-card__info"
              >
                <div
                  className="place-card__price-wrapper"
                >
                  <div
                    className="place-card__price"
                  >
                    <b
                      className="place-card__price-value"
                    >
                      €
                      120
                    </b>
                    <span
                      className="place-card__price-text"
                    >
                      / night
                    </span>
                  </div>
                  <button
                    className="place-card__bookmark-button button place-card__bookmark-button--active"
                    onClick={[Function]}
                    type="button"
                  >
                    <svg
                      className="place-card__bookmark-icon"
                      height="19"
                      width="18"
                    >
                      <use
                        xlinkHref="#icon-bookmark"
                      />
                    </svg>
                    <span
                      className="visually-hidden"
                    >
                      To bookmarks
                    </span>
                  </button>
                </div>
                <div
                  className="place-card__rating rating"
                >
                  <div
                    className="place-card__stars rating__stars"
                  >
                    <span
                      style={
                        Object {
                          "width": "100%",
                        }
                      }
                    />
                    <span
                      className="visually-hidden"
                    >
                      Rating
                    </span>
                  </div>
                </div>
                <h2
                  className="place-card__name"
                >
                  <a
                    href="/offer/2"
                    onClick={[Function]}
                  >
                    Beautiful & luxurious studio at great location
                  </a>
                </h2>
                <p
                  className="place-card__type"
                >
                  apartment
                </p>
              </div>
            </article>
            <article
              className=" cities__place-card place-card"
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
            >
              <div
                className="cities__image-wrapper place-card__image-wrapper"
              >
                <a
                  href="/offer/1"
                  onClick={[Function]}
                >
                  <img
                    alt="Place image"
                    className="place-card__image"
                    height="200"
                    src="img/1.png"
                    width="260"
                  />
                </a>
              </div>
              <div
                className="cities__card-info place-card__info"
              >
                <div
                  className="place-card__price-wrapper"
                >
                  <div
                    className="place-card__price"
                  >
                    <b
                      className="place-card__price-value"
                    >
                      €
                      120
                    </b>
                    <span
                      className="place-card__price-text"
                    >
                      / night
                    </span>
                  </div>
                  <button
                    className="place-card__bookmark-button button"
                    onClick={[Function]}
                    type="button"
                  >
                    <svg
                      className="place-card__bookmark-icon"
                      height="19"
                      width="18"
                    >
                      <use
                        xlinkHref="#icon-bookmark"
                      />
                    </svg>
                    <span
                      className="visually-hidden"
                    >
                      To bookmarks
                    </span>
                  </button>
                </div>
                <div
                  className="place-card__rating rating"
                >
                  <div
                    className="place-card__stars rating__stars"
                  >
                    <span
                      style={
                        Object {
                          "width": "100%",
                        }
                      }
                    />
                    <span
                      className="visually-hidden"
                    >
                      Rating
                    </span>
                  </div>
                </div>
                <h2
                  className="place-card__name"
                >
                  <a
                    href="/offer/1"
                    onClick={[Function]}
                  >
                    Beautiful & luxurious studio at great location
                  </a>
                </h2>
                <p
                  className="place-card__type"
                >
                  apartment
                </p>
              </div>
            </article>
          </div>
        </section>
        <div
          className="cities__right-section"
        >
          <Map
            activeScrollWheelZoom={true}
            activeZoomControl={false}
            offers={
              Array [
                Object {
                  "bedrooms": 3,
                  "city": Object {
                    "location": Object {
                      "latitude": 52.370216,
                      "longitude": 4.895168,
                      "zoom": 10,
                    },
                    "name": "Amsterdam",
                  },
                  "description": "A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.",
                  "goods": Array [
                    "Heating",
                    "Kitchen",
                    "Cable TV",
                    "Washing machine",
                    "Coffee machine",
                    "Dishwasher",
                  ],
                  "host": Object {
                    "avatar_url": "img/1.png",
                    "id": 7,
                    "is_pro": true,
                    "name": "Angelina",
                  },
                  "id": 1,
                  "images": Array [
                    "img/1.png",
                    "img/2.png",
                  ],
                  "is_favorite": false,
                  "is_premium": false,
                  "location": Object {
                    "latitude": 52.35514938496378,
                    "longitude": 4.673877537499948,
                    "zoom": 8,
                  },
                  "max_adults": 4,
                  "preview_image": "img/1.png",
                  "price": 120,
                  "rating": 4.8,
                  "title": "Beautiful & luxurious studio at great location",
                  "type": "apartment",
                },
                Object {
                  "bedrooms": 3,
                  "city": Object {
                    "location": Object {
                      "latitude": 52.370216,
                      "longitude": 4.895168,
                      "zoom": 10,
                    },
                    "name": "Amsterdam",
                  },
                  "description": "A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.",
                  "goods": Array [
                    "Heating",
                    "Kitchen",
                    "Cable TV",
                    "Washing machine",
                    "Coffee machine",
                    "Dishwasher",
                  ],
                  "host": Object {
                    "avatar_url": "img/1.png",
                    "id": 6,
                    "is_pro": true,
                    "name": "Angelina",
                  },
                  "id": 2,
                  "images": Array [
                    "img/1.png",
                    "img/2.png",
                  ],
                  "is_favorite": true,
                  "is_premium": true,
                  "location": Object {
                    "latitude": 52.35514938496378,
                    "longitude": 4.673877537499948,
                    "zoom": 8,
                  },
                  "max_adults": 4,
                  "preview_image": "img/1.png",
                  "price": 120,
                  "rating": 4.8,
                  "title": "Beautiful & luxurious studio at great location",
                  "type": "apartment",
                },
                Object {
                  "bedrooms": 2,
                  "city": Object {
                    "location": Object {
                      "latitude": 52.370216,
                      "longitude": 4.895168,
                      "zoom": 10,
                    },
                    "name": "Amsterdam",
                  },
                  "description": "A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.",
                  "goods": Array [
                    "Heating",
                    "Kitchen",
                    "Cable TV",
                    "Washing machine",
                    "Coffee machine",
                    "Dishwasher",
                  ],
                  "host": Object {
                    "avatar_url": "img/1.png",
                    "id": 2,
                    "is_pro": false,
                    "name": "Angelina",
                  },
                  "id": 10,
                  "images": Array [
                    "img/1.png",
                    "img/2.png",
                  ],
                  "is_favorite": false,
                  "is_premium": false,
                  "location": Object {
                    "latitude": 52.35514938496378,
                    "longitude": 4.673877537499948,
                    "zoom": 8,
                  },
                  "max_adults": 4,
                  "preview_image": "img/1.png",
                  "price": 120,
                  "rating": 2,
                  "title": "Beautiful & luxurious studio at great location",
                  "type": "apartment",
                },
                Object {
                  "bedrooms": 3,
                  "city": Object {
                    "location": Object {
                      "latitude": 52.370216,
                      "longitude": 4.895168,
                      "zoom": 10,
                    },
                    "name": "Amsterdam",
                  },
                  "description": "A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.",
                  "goods": Array [
                    "Heating",
                    "Kitchen",
                    "Cable TV",
                    "Washing machine",
                    "Coffee machine",
                    "Dishwasher",
                  ],
                  "host": Object {
                    "avatar_url": "img/1.png",
                    "id": 5,
                    "is_pro": false,
                    "name": "Angelina",
                  },
                  "id": 4,
                  "images": Array [
                    "img/1.png",
                    "img/2.png",
                  ],
                  "is_favorite": true,
                  "is_premium": true,
                  "location": Object {
                    "latitude": 52.35514938496378,
                    "longitude": 4.673877537499948,
                    "zoom": 8,
                  },
                  "max_adults": 4,
                  "preview_image": "img/1.png",
                  "price": 120,
                  "rating": 3,
                  "title": "Beautiful & luxurious studio at great location",
                  "type": "apartment",
                },
              ]
            }
          />
        </div>
      </div>
    </div>
  </main>
</div>
`;

exports[`MainPage render should render without offers 1`] = `
<div
  className="page page--gray page--main"
>
  <header
    className="header"
  >
    <div
      className="container"
    >
      <div
        className="header__wrapper"
      >
        <div
          className="header__left"
        >
          <a
            className="header__logo-link"
            href="/"
            onClick={[Function]}
          >
            <img
              alt="6 cities logo"
              className="header__logo"
              height="41"
              src="img/logo.svg"
              width="81"
            />
          </a>
        </div>
        <nav
          className="header__nav"
        >
          <ul
            className="header__nav-list"
          >
            <li
              className="header__nav-item user"
            >
              <a
                className="header__nav-link header__nav-link--profile"
                href="/login"
                onClick={[Function]}
              >
                <div
                  className="header__avatar-wrapper user__avatar-wrapper"
                />
                <span
                  className="header__user-name user__name"
                >
                  Sing in
                </span>
              </a>
            </li>
          </ul>
        </nav>
      </div>
    </div>
  </header>
  <main
    className="page__main undefined"
  >
    <h1
      className="visually-hidden"
    >
      Cities
    </h1>
    <div
      className="tabs"
    >
      <section
        className="locations container"
      >
        <ul
          className="locations__list tabs__list"
        >
          <li
            className="locations__item"
            onClick={[Function]}
          >
            <a
              className="locations__item-link tabs__item tabs__item--active"
            >
              <span>
                Paris
              </span>
            </a>
          </li>
          <li
            className="locations__item"
            onClick={[Function]}
          >
            <a
              className="locations__item-link tabs__item"
            >
              <span>
                Cologne
              </span>
            </a>
          </li>
          <li
            className="locations__item"
            onClick={[Function]}
          >
            <a
              className="locations__item-link tabs__item"
            >
              <span>
                Brussels
              </span>
            </a>
          </li>
          <li
            className="locations__item"
            onClick={[Function]}
          >
            <a
              className="locations__item-link tabs__item"
            >
              <span>
                Amsterdam
              </span>
            </a>
          </li>
          <li
            className="locations__item"
            onClick={[Function]}
          >
            <a
              className="locations__item-link tabs__item"
            >
              <span>
                Hamburg
              </span>
            </a>
          </li>
          <li
            className="locations__item"
            onClick={[Function]}
          >
            <a
              className="locations__item-link tabs__item"
            >
              <span>
                Dusseldorf
              </span>
            </a>
          </li>
        </ul>
      </section>
    </div>
    <div
      className="cities"
    >
      <div
        className="cities__places-container cities__places-container--empty container"
      >
        <section
          className="cities__no-places"
        >
          <div
            className="cities__status-wrapper tabs__content"
          >
            <b
              className="cities__status"
            >
              No places to stay available
            </b>
            <p
              className="cities__status-description"
            >
              We could not find any property available at the moment in 
              Paris
            </p>
          </div>
        </section>
        <div
          className="cities__right-section"
        />
      </div>
    </div>
  </main>
</div>
`;
